### PR TITLE
cmd/derpprobe,prober: add ability to continuously probe bandwidth and…

### PR DESCRIPTION
**Pull request by oxtoacart**
… track queuing delay

Setting the bw-interval to a negative value enables continuous bandwidth probing. When probing bandwidth continuously, the prober also tracks packet transit times, which can be used as a measure of queuing delay.

Updates tailscale/corp#24522

I tested this by running against a local derper and manually observing my `/metrics` endpoint. With a 64KB packet size, I was seeing 2 Gbps throughput and a queuing delay usually under 2ms.

Here is an example of the metrics as they would be exported to Prometheus.

```
# HELP derpprobe_derp_bw_probe_continuous_bytes_total Total data transferred
# TYPE derpprobe_derp_bw_probe_continuous_bytes_total counter
derpprobe_derp_bw_probe_continuous_bytes_total{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1"} 6.8105856e+10
# HELP derpprobe_derp_bw_probe_continuous_queuing_delays_seconds Distribution of queuing delays
# TYPE derpprobe_derp_bw_probe_continuous_queuing_delays_seconds histogram
derpprobe_derp_bw_probe_continuous_queuing_delays_seconds_bucket{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1",le="0.0001"} 42791
derpprobe_derp_bw_probe_continuous_queuing_delays_seconds_bucket{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1",le="0.0002"} 730496
derpprobe_derp_bw_probe_continuous_queuing_delays_seconds_bucket{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1",le="0.0005"} 287610
derpprobe_derp_bw_probe_continuous_queuing_delays_seconds_bucket{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1",le="+Inf"} 1.064154e+06
derpprobe_derp_bw_probe_continuous_queuing_delays_seconds_sum{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1"} 188.88120006199648
derpprobe_derp_bw_probe_continuous_queuing_delays_seconds_count{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1"} 1.064154e+06
# HELP derpprobe_derp_bw_probe_continuous_size_bytes Payload size of the bandwidth prober
# TYPE derpprobe_derp_bw_probe_continuous_size_bytes gauge
derpprobe_derp_bw_probe_continuous_size_bytes{class="derp_bw",derp_path="single",hostname="localhost",name="derp/dev/1f/1f/bw",region="dev",region_id="1"} 64000
```

### CODE REVIEW

File: derp/derphttp/derphttp_client.go [hunk 1]
```diff
@@ -757,6 +757,9 @@ func (c *Client) dialNode(ctx context.Context, n *tailcfg.DERPNode) (net.Conn, e
 			}
 			dst := cmp.Or(dstPrimary, n.HostName)
 			port := "443"
+			if !c.useHTTPS() {
```
Comment (oxtoacart) [inline derp/derphttp/derphttp_client.go:760]: Drive-by: when testing locally, using a `-dev` mode derper did not work we didn't implement the logic that [this](https://github.com/tailscale/tailscale/pull/14122/files#diff-99911aec55c961ef16b36e90b0f5790bae75921c381504b4f7f09f9ea7c1aa09L245) comment says we have.

File: derp/derphttp/derphttp_client.go [hunk 2]
```diff
@@ -748,7 +748,7 @@ func (c *Client) dialNode(ctx context.Context, n *tailcfg.DERPNode) (net.Conn, e
 				select {
 				case <-ctx.Done():
 					// Either user canceled original context,
-					// it timed out, or the v6 dial succeeded.
+					// if timed out, or the v6 dial succeeded.
```
Comment (oxtoacart) [inline derp/derphttp/derphttp_client.go:751]: Drive-by spelling

File: prober/derp.go [hunk 3]
```diff
@@ -196,12 +201,12 @@ func (d *derpProber) probeMapFn(ctx context.Context) error {
 					}
 				}
 
-				if d.bwInterval > 0 && d.bwProbeSize > 0 {
-					n := fmt.Sprintf("derp/%s/%s/%s/bw", region.RegionCode, server.Name, to.Name)
+				if d.bwInterval != 0 && d.bwProbeSize > 0 {
+					n := fmt.Sprintf("derp/%s/%s/%s/bw/%d", region.RegionCode, server.Name, to.Name, d.bwProbeSize)
```
Comment (oxtoacart) [inline prober/derp.go:205]: I'm including the probe size in the name for clarity. This way, if we wanted to run another probe using different packet sizes, we could look at the resulting metrics separately.

File: prober/derp.go [hunk 4]
```diff
@@ -597,18 +751,22 @@ func newConn(ctx context.Context, dm *tailcfg.DERPMap, n *tailcfg.DERPNode, isPr
...
-	}
-	if len(cs.PeerCertificates) == 0 {
-		dc.Close()
-		return nil, errors.New("no peer certificates")
-	}
-	if cs.ServerName != n.HostName {
-		dc.Close()
-		return nil, fmt.Errorf("TLS server name %q != derp hostname %q", cs.ServerName, n.HostName)
+
+	// Only verify TLS state if this is a prober.
```
Comment (oxtoacart) [inline prober/derp.go:755]: Drive-by, when testing locally with a `-dev` mode derper, this failed because we're not using TLS. The `TLS` prober does its own checking, and the `Mesh` prober passes `isProber` here, so it will still do this TLS check.

**Review by knyar — COMMENTED** [review]
I hope you're ok with a few drive-by comments. I am not involved in this project, so don't feel strongly about any of these.

File: prober/derp.go (hunk 3 above)
Comment (oxtoacart) [reply prober/derp.go:205]: Would a label achieve the same thing, or is metric value more useful?

File: prober/prober.go [hunk 5]
```diff
@@ -288,6 +288,22 @@ func (p *Probe) loop() {
 		return
 	}
 
+	if p.interval < 0 {
+		// Probe function is going to run continuously.
```
[the remark this answers is not in the document]
Comment (oxtoacart) [reply prober/prober.go:292]: Yeah, I wondered about that too. Let me give it a go and see how it looks.

File: prober/derp.go (hunk 3 above)
Comment (oxtoacart) [reply prober/derp.go]: Done, I'm calling it `derp_bw_probe_continuous_size_bytes`.

File: prober/derp.go [hunk 6]
```diff
@@ -271,6 +284,75 @@ func (d *derpProber) probeBandwidth(from, to string, size int64) ProbeClass {
...
+			return derpProbeBandwidthContinuous(ctx, d.lastDERPMap, fromN, toN, size, &bytesTransferred, &transferTime, &qdh)
+		},
+		Class:  "derp_bw",
+		Labels: Labels{"derp_path": derpPath},
+		Metrics: func(l prometheus.Labels) []prometheus.Metric {
+			qdh.mx.Lock()
+			result := []prometheus.Metric{
+				prometheus.MustNewConstMetric(prometheus.NewDesc("derp_bw_probe_continuous_bytes_total", "Total data transferred", nil, l), prometheus.CounterValue, float64(bytesTransferred.Value())),
+				prometheus.MustNewConstMetric(prometheus.NewDesc("derp_bw_probe_continuous_transfer_time_seconds_total", "Time it took to transfer data", nil, l), prometheus.CounterValue, transferTime.Value()),
+				prometheus.MustNewConstHistogram(prometheus.NewDesc("derp_bw_probe_continuous_queuing_delays_milliseconds", "Distribution of queuing delays", nil, l), qdh.count, qdh.sum, qdh.buckets),
```
[the remark this answers is not in the document]
Comment (oxtoacart) [reply prober/derp.go]: Done.

**Comment by oxtoacart** [comment]
> I hope you're ok with a few drive-by comments. I am not involved in this project, so don't feel strongly about any of these.

I'm more than okay with it, I appreciate it very much, thanks!

File: prober/prober.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (oxtoacart) [reply prober/prober.go:292]: Thinking about it some more, the library does give us more than that. It keeps the DERPmap updated and it provides us with properly configured DERP clients.

[the remark this answers is not in the document]
Comment (oxtoacart) [reply prober/prober.go:292]: I'm going to noodle on making continuous probing a first-class concept in the library and see how that looks.

[the remark this answers is not in the document]
Comment (oxtoacart) [reply prober/prober.go:292]: So, I think there's value to showing continuous probes on the web page. I'm revising the dashboard to look like the below. The key changes are:

- Show start and end times
- Show interval as "continuous" for continuous probes
- Don't show status history for continuous probes
- Don't show latencies for continuous probes

image

image

[the remark this answers is not in the document]
Comment (knyar) [reply prober/prober.go:292]: I think these changes make continuous probes less awkward, thank you. You might also want to return a 4xx error in [RunHandler](https://github.com/tailscale/tailscale/blob/main/prober/prober.go/#L461) for the continuous probe.

File: prober/derp.go [hunk 7]
```diff
@@ -577,6 +692,88 @@ func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc
...
+				sentTime := <-sentTimes
+				if v.Source != fromc.SelfPublicKey() {
+					recvc <- fmt.Errorf("got data packet from unexpected source, %v", v.Source)
+					return
+				}
+				if !bytes.Equal(v.Data, pkt) {
+					recvc <- fmt.Errorf("unexpected data packet")
+					return
+				}
+				qdh.add(now.Sub(sentTime))
```
[the remark this answers is not in the document]
Comment (oxtoacart) [reply prober/derp.go:750]: Did I misread [this](https://github.com/tailscale/tailscale/blob/main/prober/derp.go#L547)? It looked to me like the existing probe expects packets to arrive in the order they were sent.

[the remark this answers is not in the document]
Comment (jwhited) [reply prober/derp.go:750]: Same assumption is incorrect there. One of the reasons it mostly works is the fixed (30 packets) flow control being applied.

[the remark this answers is not in the document]
Comment (oxtoacart) [reply prober/derp.go]: Haha, okay. I've

1. Added a comment to the existing code that warns about this behavior

2. For the continuous probe, I'm marshaling the monotonic time into the packet headers. I had to do some reflection for this because the monotonic time data isn't exposed by the public time.Time methods. AFAICT, it's safe to reuse the single byte slice for the packet as each call to `Send()` ends up copying it anyway.

**Comment by oxtoacart** [comment]
We're going to hold off on continuous bandwidth probing for now.

**Comment by oxtoacart** [comment]
Repurposing to handle queuing delay probes

**Comment by jwhited** [comment]
> Repurposing to handle queuing delay probes

open a new PR, branch can be re-used though

**Comment by oxtoacart** [comment]
Closing in favor of #14405
